sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -151,6 +164,30 @@ abstract class Expression extends TreeNode[Expression] {
   }
 
   /**
+   * Returns the hash for this expression. Expressions that compute the same result, even if
+   * they differ cosmetically should return the same hash.
+   */
+  def semanticHash() : Int = {
```

Comment:
And there is more discussion on the `semanticEquals` can be found at #6587, even, I don't think we need the `semanticEquals` if we changed the implementation of `AttributeReference.equals`, as it does make lots of code complicated like this one, by using the `semanticEquals`, other than `equals` or `==`.